backends/intel_gpu/kernels/matmul_kernel.cc

```diff
@@ -0,0 +1,818 @@
+      auto N = transpose_x ? x_dims[x_ndim - 2] : x_dims[x_ndim - 1];
+      GEMM(false,
+           !transpose_y,
+           M,
+           K,
+           N,
+           out_grad_data,
+           y_data,
+           dx_data,
+           transpose_x);
+    }
+    // dy = x' * dout
+    if (dy) {
+      auto M = transpose_y ? y_dims[y_ndim - 1] : y_dims[y_ndim - 2];
+      auto K = transpose_x ? x_dims[x_ndim - 1] : x_dims[x_ndim - 2];
+      auto N = transpose_y ? y_dims[y_ndim - 2] : y_dims[y_ndim - 1];
+      GEMM(!transpose_x,
+           false,
+           M,
+           K,
+           N,
+           x_data,
+           out_grad_data,
+           dy_data,
+           transpose_y);
+    }
+    return;
+  } else if (x_ndim == 2) {
+    // dx = dout * y'
+    if (dx) {
+      auto M = transpose_x ? x_dims[x_ndim - 1] : x_dims[x_ndim - 2];
+      auto K = transpose_y ? y_dims[y_ndim - 2] : y_dims[y_ndim - 1];
+      auto N = transpose_x ? x_dims[x_ndim - 2] : x_dims[x_ndim - 1];
+      BatchedGEMM(false,
+                  !transpose_y,
+                  M,
+                  K,
+                  N,
+                  out_grad_data,
+                  y_data,
+                  dx_data,
+                  y_dims[0],
+                  false,
+                  transpose_x,
+                  true,
+                  true);
+    }
+    // dy = x' * dout
+    if (dy) {
+      auto M = transpose_y ? y_dims[y_ndim - 1] : y_dims[y_ndim - 2];
+      auto K = transpose_x ? x_dims[x_ndim - 1] : x_dims[x_ndim - 2];
+      auto N = transpose_y ? y_dims[y_ndim - 2] : y_dims[y_ndim - 1];
+      BatchedGEMM(!transpose_x,
+                  false,
+                  M,
+                  K,
+                  N,
+                  x_data,
+                  out_grad_data,
+                  dy_data,
+                  y_dims[0],
+                  false,
+                  transpose_y);
+    }
+    return;
+  } else if (y_ndim == 2) {
+    // dx = dout * y'
+    if (dx) {
+      auto M = transpose_x ? x_dims[x_ndim - 1] : x_dims[x_ndim - 2];
+      auto K = transpose_y ? y_dims[y_ndim - 2] : y_dims[y_ndim - 1];
+      auto N = transpose_x ? x_dims[x_ndim - 2] : x_dims[x_ndim - 1];
+      BatchedGEMM(false,
+                  !transpose_y,
+                  M,
+                  K,
+                  N,
+                  out_grad_data,
+                  y_data,
+                  dx_data,
+                  x_dims[0],
+                  true,
+                  transpose_x);
+    }
+    // dy = x' * dout
+    if (dy) {
+      auto M = transpose_y ? y_dims[y_ndim - 1] : y_dims[y_ndim - 2];
+      auto K = transpose_x ? x_dims[x_ndim - 1] : x_dims[x_ndim - 2];
+      auto N = transpose_y ? y_dims[y_ndim - 2] : y_dims[y_ndim - 1];
+      GEMM(!transpose_x,
+           false,
+           M,
+           K * x_dims[0],
+           N,
+           x_data,
+           out_grad_data,
+           dy_data,
+           transpose_y);
+    }
+    return;
+  }
+}
```

Comment:
all the above code is needless in intel_gpu, please remove and do not copy kernel code from custom cpu as they will not be used by device kernel code.

Reply:
done